### ISSUE TITLE
Misleading compiler errors from match expressions inside loops with branches of infallible

### ISSUE THREAD

**Issue by simvux** [rust-lang/rust#83931]
Given the following code: 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=83112e42531e00a584b38696d5997e54

```rust
enum MyEnum {
    First,
    Second,
    Third,
    Fourth,
}

fn main() {
    let a = MyEnum::First;

    for _ in 0..1 {
        match a {
            First => Err(()),
            Second => Ok(()),
            Third => {
                todo!()
            }
            _ => Ok(()),
        }
    }
}
```

The current output is:

```
error[E0308]: `match` arms have incompatible types
  --> src/main.rs:16:17
   |
12 | /         match a {
13 | |             First => Err(()),
   | |                      ------- this is found to be of type `std::result::Result<(), ()>`
14 | |             Second => Ok(()),
   | |                       ------ this is found to be of type `std::result::Result<(), ()>`
15 | |             Third => {
16 | |                 todo!()
   | |                 ^^^^^^^ expected enum `std::result::Result`, found `()`
17 | |             }
18 | |             _ => Ok(()),
19 | |         }
   | |_________- `match` arms have incompatible types
   |
   = note:   expected enum `std::result::Result<(), ()>`
           found unit type `()`
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: try using a variant of the expected enum
   |
15 |             Third => Ok({
16 |                 todo!()
17 |             })
   |
15 |             Third => Err({
16 |                 todo!()
17 |             })
   |

error: aborting due to previous error
```

I suspect the infallible branch is instead assigned the type `()` due to the expected type from the for loop. Causing the misleading help and arrows. 

Changing `todo!()` into `Ok(())` produces the expected error output of:
```
error[E0308]: mismatched types
  --> src/main.rs:12:9
   |
12 | /         match a {
13 | |             First => Err(()),
14 | |             Second => Ok(()),
15 | |             Third => Ok(()),
16 | |             _ => Ok(()),
17 | |         }
   | |         ^- help: try adding a semicolon: `;`
   | |_________|
   |           expected `()`, found enum `std::result::Result`
   |
   = note: expected unit type `()`
                   found enum `std::result::Result<(), ()>`

error: aborting due to previous error
```

**Comment by eggyal** [comment]
It's also interesting that the problem does not arise [when the infallible arm is not a block expression](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=607661eb8a5f80b8b173a4bd6dc96413), which suggests the type of the block is also not being correctly determined.

**Comment by yaymukund** [comment]
I believe I ran into this using continue. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3a6a44f02838b0bfb499cbbc09b27604

> It's also interesting that the problem does not arise [when the never arm is not a block expression](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=607661eb8a5f80b8b173a4bd6dc96413), which suggests the type of the block is also not being correctly determined.

I don't think it matters anymore whether the match arm is a block expression. Your playground behaves the same with or without braces `{ }`.

**Comment by conradludgate** [comment]
This issue can be reduced down to

```rust
fn main() {
    let a = Some(());
    match a {
        Some(()) => Some(()),
        _ => { todo!() }
    }
}
```
